Jenkinsfile: build Creator-feed from Creator SDK

nikhil-zinjurde-imgtec opened the pull request:
This closes CreatorDev/openwrt#187

Signed-off-by: Nikhil Zinjurde <nikhil.zinjurde@imgtec.com>

nikhil-zinjurde-imgtec committed:
Jenkinsfile: build Creator-feed from Creator SDK

Signed-off-by: Nikhil Zinjurde <nikhil.zinjurde@imgtec.com>

MattAtkinson reviewed (approved):
shell commands look good

Ham22 commented on Jenkinsfile:
```diff
@@ -1,15 +1,72 @@
+        docker_image = docker.image "imgtec/creator-builder:latest"
+    }
+    docker_image.inside {
+//      deleteDir() // Wipe the workspace so we are building completely clean
```
This is what clean checkout below does?

Ham22 commented on Jenkinsfile:
```diff
@@ -1,15 +1,72 @@
+            ])
+
+            // Download and extract Creator SDK
+            sh "wget -O ${WORKSPACE}/sdk.tar.bz2 \
```
These commands can be piped together to avoid the massive temporary file

Ham22 commented on Jenkinsfile:
```diff
@@ -1,15 +1,72 @@
+
+            // Download and extract Creator SDK
+            sh "wget -O ${WORKSPACE}/sdk.tar.bz2 \
+                http://downloads.creatordev.io/creator-openwrt/v1/pistachio/marduk/creator-sdk-1.tar.bz2 && \
```
I would expect this location to be a property so I can override the default SDK to build with?

Ham22 commented on Jenkinsfile:
```diff
@@ -1,15 +1,72 @@
+        }
+        stage('Upload') {
+            archiveArtifacts '**/bin/pistachio/creator/*'
+            dir('creator-sdk') {
```
Why inside this dir and not top level (also dir doesn't work inside docker)?

Ham22 commented on Jenkinsfile:
```diff
@@ -1,15 +1,72 @@
+                tar -xf ${WORKSPACE}/sdk.tar.bz2"
+
+            // Enable signed package indexing
+            sh "echo '\nconfig SIGNED_PACKAGES\n\tbool\n\tdefault y\n' >> ${WORKSPACE}/creator-sdk/Config.in"
```
There is no need for `${workspace}` in all these commands as its all relatvie to where you currently are anyway

Ham22 commented on Jenkinsfile:
```diff
@@ -1,15 +1,72 @@
+
+            // Update and install all feeds
+            sh "cd ${WORKSPACE}/creator-sdk && cat Config.in feeds.conf.default"
+            sh "cd ${WORKSPACE}/creator-sdk && scripts/feeds update -a && scripts/feeds install -a -p creator"
```
When doing this manually I found that dependencies were not being installed properly with `-p creator`?

Ham22 commented on Jenkinsfile:
```diff
@@ -1,15 +1,72 @@
+        stage('Build') {
+            // Attempt to build quickly and reliably
+            try {
+                sh "cd ${WORKSPACE}/creator-sdk && make -j4 V=s"
```
When building manually I found that you had to specify each package in our extra feed to be built (in dependency order) otherwise it would build far more than it needed to.

Ham22 commented on Jenkinsfile:
```diff
@@ -1,15 +1,82 @@
+
+def creatorPackages = [
+    ['awalwm2m', 'gcc53', 'info-daemon', 'letmecreate', 'onboarding-scripts', 'provisioning-daemon', 'pyletmecreate'],
+] // TODO pickle package not added due to compilation error
```
What was the error as it works in manual builds. Note it depends on bit-bang-io which doesn't seem to be in this list either?

Ham22 commented on Jenkinsfile:
```diff
@@ -1,15 +1,82 @@
+        }
+        stage('Build') {
+            // Attempt to build quickly and reliably
+            // TODO build specific package only based on PR or commit
```
TODOs in comments like this will get lost and not ever prioritised. I think best solution is to create a ticket and point the comment to bug number instead?

Ham22 commented on Jenkinsfile:
```diff
@@ -1,15 +1,82 @@
+            ])
+
+            // Download and extract Creator SDK
+            sh "mkdir creator-sdk && wget -qO- ${params.SDK_TARBALL} | tar -xvf -C creator-sdk --strip-components 1"
```
This won't work? You are piping it so shouldn't give a `-f` flag also if its bz2 you need to pass `-j` as it can't work it out when piped.

Ham22 commented on Jenkinsfile:
```diff
@@ -1,15 +1,82 @@
+    }
+    docker_image.inside {
+        stage('Prepare') {
+            checkout([$class: 'GitSCM',
```
Seeing as the feed is built inside the SDK then perhaps logically it would make more sense to switch the flow around a bit:
1. download the sdk to the top level of workspace (no need to create sub-dir)
2. checkout feed to sub-dir
3. point feed conf to local feed
4. build

Then there is no need for any change directory commands :D

Ham22 commented on Jenkinsfile:
```diff
@@ -1,15 +1,82 @@
+ */
+
+def creatorPackages = [
+    ['awalwm2m', 'gcc53', 'info-daemon', 'letmecreate', 'onboarding-scripts', 'provisioning-daemon', 'pyletmecreate'],
```
Why 2d?

Ham22 commented on Jenkinsfile:
```diff
@@ -1,15 +1,82 @@
+                    echo 'Parallel build failed, attempting to continue in single threaded mode'
+                    sh "cd creator-sdk && make package/${package}/compile -j1 V=s"
+                }
+            sh "cd creator-sdk && make package/index -j1 V=s"
```
Indentation

Ham22 commented on Jenkinsfile:
```diff
@@ -1,15 +1,82 @@
+        stage('Build') {
+            // Attempt to build quickly and reliably
+            // TODO build specific package only based on PR or commit
+            for (package in creatorPackages) {
```
package is a keyword isn't it?

nikhil-zinjurde-imgtec commented:
Creator packages generated on CI along with signed index for testing, here - http://jenkins.creatordev.io/job/CreatorDev/job/Creator-feed/view/Pull%20Requests/job/PR-31/13/artifact/bin/pistachio/packages/creator/

nikhil-zinjurde-imgtec commented:
Creator packages generated on CI along with signed index, here - http://jenkins.creatordev.io/job/CreatorDev/job/Creator-feed/view/Pull%20Requests/job/PR-31/13/artifact/bin/pistachio/packages/creator/

nikhil-zinjurde-imgtec commented:
## Verification passed
* Added creator package index

        root@OpenWrt:/# cat /etc/opkg/distfeeds.conf
        src/gz ci40_base http://downloads.creatordev.io/openwrt/latest/pistachio/marduk/packages/base
        src/gz ci40_kernel http://downloads.creatordev.io/openwrt/latest/pistachio/marduk/packages/kernel
        src/gz ci40_ci40 http://downloads.creatordev.io/openwrt/latest/pistachio/marduk/packages/ci40
        src/gz ci40_telephony http://downloads.creatordev.io/openwrt/latest/pistachio/marduk/packages/telephony
        src/gz ci40_packages http://downloads.creatordev.io/openwrt/latest/pistachio/marduk/packages/packages
        src/gz ci40_routing http://downloads.creatordev.io/openwrt/latest/pistachio/marduk/packages/routing
        src/gz ci40_luci http://downloads.creatordev.io/openwrt/latest/pistachio/marduk/packages/luci
        src/gz ci40_management http://downloads.creatordev.io/openwrt/latest/pistachio/marduk/packages/management
        src/gz ci40_creator http://jenkins.creatordev.io/job/CreatorDev/job/Creator-feed/view/Pull%20Requests/job/PR-31/13/artifact/bin/pistachio/packages/creator
        # src/gz ci40_targets http://downloads.creatordev.io/openwrt/latest/pistachio/marduk/packages/targets

* Ran opkg update

        root@OpenWrt:/# opkg update
        Downloading http://downloads.creatordev.io/openwrt/latest/pistachio/marduk/packages/base/Packages.gz.
        Updated list of available packages in /var/opkg-lists/ci40_base.
        Downloading http://downloads.creatordev.io/openwrt/latest/pistachio/marduk/packages/base/Packages.sig.
        Signature check passed.
        Downloading http://downloads.creatordev.io/openwrt/latest/pistachio/marduk/packages/kernel/Packages.gz.
        Updated list of available packages in /var/opkg-lists/ci40_kernel.
        Downloading http://downloads.creatordev.io/openwrt/latest/pistachio/marduk/packages/kernel/Packages.sig.
        Signature check passed.
        Downloading http://downloads.creatordev.io/openwrt/latest/pistachio/marduk/packages/ci40/Packages.gz.
        Updated list of available packages in /var/opkg-lists/ci40_ci40.
        Downloading http://downloads.creatordev.io/openwrt/latest/pistachio/marduk/packages/ci40/Packages.sig.
        Signature check passed.
        Downloading http://downloads.creatordev.io/openwrt/latest/pistachio/marduk/packages/telephony/Packages.gz.
        Updated list of available packages in /var/opkg-lists/ci40_telephony.
        Downloading http://downloads.creatordev.io/openwrt/latest/pistachio/marduk/packages/telephony/Packages.sig.
        Signature check passed.
        Downloading http://downloads.creatordev.io/openwrt/latest/pistachio/marduk/packages/packages/Packages.gz.
        Updated list of available packages in /var/opkg-lists/ci40_packages.
        Downloading http://downloads.creatordev.io/openwrt/latest/pistachio/marduk/packages/packages/Packages.sig.
        Signature check passed.
        Downloading http://downloads.creatordev.io/openwrt/latest/pistachio/marduk/packages/routing/Packages.gz.
        Updated list of available packages in /var/opkg-lists/ci40_routing.
        Downloading http://downloads.creatordev.io/openwrt/latest/pistachio/marduk/packages/routing/Packages.sig.
        Signature check passed.
        Downloading http://downloads.creatordev.io/openwrt/latest/pistachio/marduk/packages/luci/Packages.gz.
        Updated list of available packages in /var/opkg-lists/ci40_luci.
        Downloading http://downloads.creatordev.io/openwrt/latest/pistachio/marduk/packages/luci/Packages.sig.
        Signature check passed.
        Downloading http://downloads.creatordev.io/openwrt/latest/pistachio/marduk/packages/management/Packages.gz.
        Updated list of available packages in /var/opkg-lists/ci40_management.
        Downloading http://downloads.creatordev.io/openwrt/latest/pistachio/marduk/packages/management/Packages.sig.
        Signature check passed.
        Downloading http://jenkins.creatordev.io/job/CreatorDev/job/Creator-feed/view/Pull%20Requests/job/PR-31/13/artifact/bin/pistachio/packages/creator/Packages.gz.
        Updated list of available packages in /var/opkg-lists/ci40_creator.
        Downloading http://jenkins.creatordev.io/job/CreatorDev/job/Creator-feed/view/Pull%20Requests/job/PR-31/13/artifact/bin/pistachio/packages/creator/Packages.sig.
        Signature check passed.

* Installed awalwm2m

        root@OpenWrt:/# opkg install awalwm2m
        Installing awalwm2m (0.2.4-1) to root...
        Downloading http://jenkins.creatordev.io/job/CreatorDev/job/Creator-feed/view/Pull%20Requests/job/PR-31/13/artifact/bin/pistachio/packages/creator/awalwm2m_0.2.4-1_pistachio.ipk.
        Installing libgnutls (3.5.6-1) to root...
        Downloading http://downloads.creatordev.io/openwrt/latest/pistachio/marduk/packages/packages/libgnutls_3.5.6-1_pistachio.ipk.
        Installing libnettle (3.2-1) to root...
        Downloading http://downloads.creatordev.io/openwrt/latest/pistachio/marduk/packages/base/libnettle_3.2-1_pistachio.ipk.
        Installing libgmp (6.1.1-1) to root...
        Downloading http://downloads.creatordev.io/openwrt/latest/pistachio/marduk/packages/base/libgmp_6.1.1-1_pistachio.ipk.
        Configuring libgmp.
        Configuring libnettle.
        Configuring libgnutls.
        Configuring awalwm2m.

Ham22 reviewed (approved):
Code looks good!

Potentially the public key isn't needed?
